src/api.ts

```diff
@@ -99,34 +101,87 @@ export async function deleteImage(hash: CommitHash) {
 		l.error({ err, commitHash: hash }, 'failed to remove image');
 	}
 }
-
 export async function startContainer(commitHash: CommitHash) {
 	l.log({ commitHash }, `Starting up container for ${commitHash}`);
 	const image = getImageName(commitHash);
+
+	// do we have an existing container?
+	const existingContainer = getRunningContainerForHash( commitHash );
+	if ( existingContainer ) {
+		l.log( { commitHash, containerId: existingContainer.Id }, `Found a running container for ${commitHash}`);
+		return Promise.resolve( existingContainer );
+	}
+
+	// are we starting one already?
+	if( state.startingContainers.has( commitHash ) ) {
+		l.log( { commitHash, containerId: existingContainer.Id }, `Already starting a container for ${commitHash}`)
+		return state.startingContainers.get( commitHash );
+	}
+
+	const startPromise = _startContainer( image, commitHash );
```

Comment:
I feel like at the point where `startContainer()` calls `_startContainer()` we're probably due for a rename somewhere

Reply:
yeah that's a sad consequence of how async / await and the outer logic all intertwine. I tried to have it all inline in startContainer but couldn't figure out how to make it work.

Reply:
moved it to a nested function